horovod/runner/elastic/discovery.py

```diff
@@ -39,13 +67,50 @@ def get_event(self):
     def set_event(self):
         self._event.set()
 
+    def _in_cooldown_period(self, current_time):
+        return self._cooldown_period_end_ts > current_time
+
+
+    def _set_cooldown_period(self, current_time):
+        if self._cooldown_lower_limit == -1 or self._cooldown_upper_limit == -1:
+            return
+        self._blacklist_count += 1
+
+        def _exponential_backoff_time():
+            cooldown_delay = self._cooldown_lower_limit * (1 << self._blacklist_count) + (random.uniform(0,1) * self._cooldown_lower_limit)
+            logging.debug(f"{self._blacklist_count}:{self._cooldown_period_end_ts} cooldown_delay: {cooldown_delay}")
+            # We need to ensure that the cooldown upper limit is the upper bound of the delay
+            return max(self._cooldown_lower_limit, min(self._cooldown_upper_limit, cooldown_delay))
+
+        cooldown_delta_seconds = _exponential_backoff_time()
```

Comment:
The inner `_exponential_backoff_time` is not really needed:

```suggestion
    def _set_cooldown_period(self, current_time):
        if self._cooldown_lower_limit == -1 or self._cooldown_upper_limit == -1:
            return
        self._blacklist_count += 1

        cooldown_delay = self._cooldown_lower_limit * (1 << self._blacklist_count) + (random.uniform(0,1) * self._cooldown_lower_limit)
        logging.debug(f"{self._blacklist_count}:{self._cooldown_period_end_ts} cooldown_delay: {cooldown_delay}")
        # We need to ensure that the cooldown upper limit is the upper bound of the delay
        cooldown_delta_seconds = max(self._cooldown_lower_limit, min(self._cooldown_upper_limit, cooldown_delay))
```

Reply:
@EnricoMi I was looking for a clean way to separate it from the rest of the cooldown calculation for readability. However, I can inline it as you have suggested.

Reply:
Understood. I think two empty lines around that block should make it readable as well.